keps/sig-apps/20200529-delete-priority-annotations.md

```diff
@@ -0,0 +1,293 @@
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] (R) Graduation criteria is in place
+- [ ] (R) Production readiness review completed
+- [ ] Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This feature allows making a suggestion to the scheduler, which pod of a Deployment should be deleted first when a scale-down event happens. This can prevent session disruption in stateful applications in a trivial manner.
+
+## Motivation
+
+For some applications, it is necessary that the application can tell Kubernetes which pod can be deleted and which replica has to be protected. The reason for this is that some applications do have stateful sessions and it is not possible to put such an application into Kubernetes because of data or session corruption resulting from "random" down-scale. If the application is able to tell Kubernetes which of the replicas contains no/few/less important active sessions, this would solve many problems. This feature is non-disruptive to the default behaviour. Only if the annotation is existing, it will make a difference in deletion order.
+
+### Goals
+
+Solve issues so that they don't require application modification:
+* [45509](https://github.com/kubernetes/kubernetes/issues/45509) - Scale down a deployment by removing specific pods
+
+
+### Non-Goals
+
+Make explicit choices to select a replica that gets deleted.
+
+## Proposal
+
+The application can set the `controller.alpha.kubernetes.io/delete-priority` annotation to a pod through the Kubernetes API. When a downscale event happens, the pod with the lower priority value of the previously set annotation will be deleted first. If one pod of the Deployment has no priority annotation set, it will be treated as the lowest priority.
+
+If all pods have the same priority, there is no difference in the normal pod delete decision behaviour. The same applies if the delete priority annotation is not used at all.
+
+The delete priority annotation can be changed during operation, for example, if workload changes or a new master gets elected.
+
+### User Stories (optional)
+
+
+#### Story 1
+
+In an application environment with stateful worker (user-)sessions, it is essential to keep the user sessions alive as good as possible. In case of a scale-down event, the application has to tell the scheduler, which delete decision would have the lowest impact on existing sessions.
```

Comment:
> In an application environment with stateful worker (user-)sessions,

Isn't Deployment designed for stateless workloads in which pods are designed to be identical? The use cases here and below read like we are trying to use Deployment to tackle stateful workloads which I'm not very sure.

Reply:
Hi resouer, 
for some specific applications it is essential to keep a long-live user sessions (especially for various legacy applications).

The delete priority for replicas in a deployment is definitely an unsolved issue requested from many parties. Please see issue [#45509](https://github.com/kubernetes/kubernetes/issues/45509) and my [pull request](https://github.com/kubernetes/kubernetes/pull/87849), where I was asked to create this KEP.

Reply:
`In case of a scale-down event, the application has to tell the scheduler`

Scheduler will not be listening :) `s/the scheduler/a controller` is more meaningful

Reply:
I would say "the system"